File: src/crew/hireCrew.ts
import type { CrewCandidate } from "../data/types/crewCandidate.ts";
import type { CrewMember } from "../data/types/crewMember.ts";
import type { PlanetCrewPool } from "../data/types/planetCrewPool.ts";
import type { CrewCapacity } from "../data/types/crewCapacity.ts";
import type { Wallet } from "../data/types/wallet.ts";
import type { HireResult } from "../data/types/hireResult.ts";
import { CREW_HIRE_COST_BY_TIER, CREW_WAGE_BY_TIER } from "../data/constants/crewConfig.ts";

// Phase 4 GDD §2.3/§2.4. Necessary completion: takes the actual pool/
// capacity/wallet/existing-crew data directly rather than IDs into an
// implicit store -- same pure-function reasoning already applied to
// Agent 11's purchaseListing()/getGlobalPrice(). Also checks wallet
// sufficiency before deducting, which Agent 11's purchaseListing() never
// did (a known, separately-documented gap there) -- "deducts the cost"
// only makes sense here if there's something to deduct from.
export function hireCrew(
  candidate: CrewCandidate,
  pool: PlanetCrewPool,
  capacity: CrewCapacity,
  existingCrew: CrewMember[],
  wallet: Wallet,
  playerId: string,
  now: number = Date.now(),
): HireResult {
  if (!pool.availableHires.some((entry) => entry.id === candidate.id)) {
    return { hired: false, reason: "candidate is not in this planet's crew pool" };
  }

  const maxCrew = capacity.baseCapacity + capacity.purchasedSlots;
  if (existingCrew.length >= maxCrew) {
    return { hired: false, reason: `at crew capacity (${existingCrew.length}/${maxCrew})` };
  }

  const cost = CREW_HIRE_COST_BY_TIER.find((entry) => entry.tier === candidate.tier)?.cost;
  if (cost === undefined) {
    throw new RangeError(`no hire cost defined for tier ${candidate.tier}`);
  }
  if (wallet.credits < cost) {
    return { hired: false, reason: `insufficient funds: need ${cost}, have ${wallet.credits}` };
  }

  const wageAmount = CREW_WAGE_BY_TIER.find((entry) => entry.tier === candidate.tier)?.wage;
  if (wageAmount === undefined) {
    throw new RangeError(`no wage defined for tier ${candidate.tier}`);
  }

  const crewMember: CrewMember = {
    id: candidate.id,
    hiredByPlayerId: playerId,
    tier: candidate.tier,
    profession: candidate.profession,
    status: "idle",
    assignedCraftId: null,
    hiredAt: now,
    lastCheckedAt: now,
    wageAmount,
    lastPaidAt: now,
  };

  return {
    hired: true,
    crewMember,
    updatedPool: {
      ...pool,
      availableHires: pool.availableHires.filter((entry) => entry.id !== candidate.id),
    },
    updatedWallet: { ...wallet, credits: wallet.credits - cost },
  };
}

File: unity/ProfitableCore/Simulation/HireCrew.cs
using Profitable.Core.Constants;
using Profitable.Core.Schema;

namespace Profitable.Core.Simulation;

// Ports src/crew/hireCrew.ts. Takes the actual pool/capacity/wallet/
// existing-crew data directly rather than IDs into an implicit store --
// same pure-function reasoning already applied to PurchaseListing/
// GetGlobalPrice. Also checks wallet sufficiency before deducting, which
// PurchaseListing never did (a known, separately-documented gap there).
public static class HireCrewSimulation
{
    public static HireResult HireCrew(
        CrewCandidate candidate,
        PlanetCrewPool pool,
        CrewCapacity capacity,
        IReadOnlyList<CrewMember> existingCrew,
        Wallet wallet,
        string playerId,
        long nowMs)
    {
        if (!pool.AvailableHires.Any(entry => entry.Id == candidate.Id))
        {
            return new HireRejected { Reason = "candidate is not in this planet's crew pool" };
        }

        var maxCrew = capacity.BaseCapacity + capacity.PurchasedSlots;
        if (existingCrew.Count >= maxCrew)
        {
            return new HireRejected { Reason = $"at crew capacity ({existingCrew.Count}/{maxCrew})" };
        }

        if (!CrewConfig.CrewHireCostByTier.TryGetValue(candidate.Tier, out var cost))
        {
            throw new InvalidOperationException($"no hire cost defined for tier {candidate.Tier}");
        }
        if (wallet.Credits < cost)
        {
            return new HireRejected { Reason = $"insufficient funds: need {cost}, have {wallet.Credits}" };
        }

        if (!CrewConfig.CrewWageByTier.TryGetValue(candidate.Tier, out var wageAmount))
        {
            throw new InvalidOperationException($"no wage defined for tier {candidate.Tier}");
        }

        var crewMember = new CrewMember
        {
            Id = candidate.Id,
            HiredByPlayerId = playerId,
            Tier = candidate.Tier,
            Profession = candidate.Profession,
            Status = CrewStatus.Idle,
            AssignedCraftId = null,
            HiredAt = nowMs,
            LastCheckedAt = nowMs,
            WageAmount = wageAmount,
            LastPaidAt = nowMs,
        };

        var updatedPool = new PlanetCrewPool
        {
            PlanetId = pool.PlanetId,
            AvailableHires = pool.AvailableHires.Where(entry => entry.Id != candidate.Id).ToList(),
            LastRefreshedAt = pool.LastRefreshedAt,
        };

        return new HireSucceeded
        {
            CrewMember = crewMember,
            UpdatedPool = updatedPool,
            UpdatedWallet = new Wallet { PlayerId = wallet.PlayerId, Credits = wallet.Credits - cost },
        };
    }
}
